pyfr/integrators/multip.py

```diff
@@ -0,0 +1,279 @@
+# -*- coding: utf-8 -*-
+
+import re
+
+from pyfr.inifile import Inifile
+from pyfr.integrators.dual.base import BaseDualIntegrator
+from pyfr.util import memoize, proxylist
+
+
+class MultiP(BaseDualIntegrator):
+    def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
+        # Get order
+        order = cfg.getint('solver', 'order')
+
+        # Get multigrid cycle
+        mgsect = 'pseudo-multigrid'
+        self.lvliters = list(cfg.getliteral(mgsect, 'pseudo-mgcycle'))
+        self.nlevels = len(self.lvliters)
+        self.level = 0
+
+        if self.nlevels > order:
+            raise ValueError('The number of multigrid levels cannot exceed '
+                             'the solution order')
+
+        # The maximum and minimum number of multigrid cycles
+        self.maxcycles = cfg.getint(mgsect, 'pseudo-mgcycles-max')
+        self.mincycles = cfg.getint(mgsect, 'pseudo-mgcycles-min')
+
+        if self.maxcycles < self.mincycles:
+            raise ValueError('The maximum number of multigrid cycles must '
+                             'be greater than or equal to the minimum')
+
+        # Multigrid pseudo-time steps
+        dtau = cfg.getfloat('solver-time-integrator', 'pseudo-dt')
+        dtaufact = cfg.getfloat(mgsect, 'pseudo-dt-fact', 1.0)
+        self.dtaus = [dtau*dtaufact**i for i in range(self.nlevels)]
+
+        # Generate multiple cfgs for the multigrid systems
+        self.mgcfg = [Inifile(cfg.tostr()) for i in range(self.nlevels)]
+        for deg in range(1, order + 1):
+            self.mgcfg[deg - 1].set('solver', 'order', order - (deg - 1))
+            for sec in cfg.sections():
+                m = re.match(r'solver-(.*)-mg-p{0}'.format(deg), sec)
+                if m:
+                    self.mgcfg[order - deg].rename_section(
+                        m.group(0), 'solver-' + m.group(1)
+                    )
+
+        super().__init__(backend, systemcls, rallocs, mesh, initsoln, cfg)
+
+        # Delete remaining elements maps from multigrid systems
+        for sys in self._mgsystem[1:]:
+            del sys.ele_map
+
+    @property
+    def _idxcurr(self):
+        return self._mgidxcurr[self.level]
+
+    @_idxcurr.setter
+    def _idxcurr(self, y):
+        self._mgidxcurr[self.level] = y
+
+    @property
+    def _regs(self):
+        return self._mgregs[self.level]
+
+    @_regs.setter
+    def _regs(self, y):
+        self._mgidxcurr[self.level] = y
+
+    def _init_prolrest_matrices(self):
+        self.restmat, self.prolmat = [0]*self.nlevels, [0]*self.nlevels
```

Comment:
Use a single dict indexed by tuple so:

self.projmats = {}

for ...:
    self.projmets[l1, l2] = ...

Cleans up a lot of code downstream, too.